File: src/frontend/network/transport/tcp/tcp_transport.py
import os
import socket
import threading
from typing import Callable, Optional

from network.transport.transport_interface import ITransport
from utils.logger import Logger


class TcpTransport(ITransport):
    """Class for TCP transport"""

    def __init__(self, host: str, port: int):
        """
        Construct a TCP transport with an existing socket.

        @param socket_fd The file descriptor of the connected socket.
        """
        self.host = host,
        self.port = port,
        self.fd = -1
        self.running = False
        self.reader_thread: Optional[threading.Thread] = None
        self.logger_ = Logger()

    def __del__(self):
        """Destructor ensures that the socket is closed."""
        self.close()

    def connect(self) -> bool:
        """
        Connect to the server via TCP.
        
        @return True if connection successful, False otherwise
        """
        try:
            # Create TCP socket
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.connect((self.host, self.port))
            
            # Get file descriptor
            self.fd = sock.fileno()
            
            # Detach the socket so it won't be closed when sock goes out of scope
            sock.detach()
            
            self.logger_.info(f"Connected to {self.host}:{self.port} via TCP")
            return True
            
        except Exception as e:
            self.logger_.error(f"TCP connection failed: {e}")
            return False

    def start(self, on_receive: Callable[[str], None]) -> None:
        """
        Start receiving data on the transport.

        @param on_receive Callback function to handle received data.
        """
        if self.fd < 0:
            self.logger_.error("Cannot start: not connected")
            return
        
        self.running = True

        def reader_loop():
            while self.running:
                try:
                    # Read from socket (max 4096 bytes)
                    data = os.read(self.fd, 4096)

                    # Connection closed or broken
                    if len(data) == 0:
                        self.logger_.info("Server closed connection")
                        self.running = False
                        break

                    payload = data.decode("utf-8")
                    on_receive(payload)

                except OSError as e:
                    if self.running:
                        self.logger_.error(f"Read error: {e}")
                    self.running = False
                    break
                except Exception as e:
                    self.logger_.info(f"Read error: {e}")
                    self.running = False
                    break

        self.reader_thread = threading.Thread(target=reader_loop, daemon=True)
        self.reader_thread.start()

    def send(self, data: str) -> None:
        """
        Send data over the transport.

        @param data The data to send.
        """
        if self.fd < 0:
            self.logger_.error("Cannot send: socket closed")
            return

        try:
            os.write(self.fd, data.encode("utf-8"))
        except OSError as e:
            self.logger_.error(f"Send error: {e}")
        except Exception as e:
            self.logger_.error(f"Unexpected send error: {e}")


    def close(self) -> None:
        """Close the TCP socket and terminate the reading loop."""
        self.running = False

        if self.fd >= 0:
            try:
                # Shutdown both directions
                sock = socket.socket(fileno=self.fd)
                sock.shutdown(socket.SHUT_RDWR)
                sock.close()
            except Exception:
                pass
            finally:
                self.fd = -1

        # Wait for reader thread to finish
        if self.reader_thread and self.reader_thread.is_alive():
            self.reader_thread.join(timeout=1.0)

        self.logger_.debug("TCP transport closed")

File: src/backend/exe/network/transport/tcp/TcpTransport.cpp
#include "TcpTransport.hpp"

#include <sys/socket.h>
#include <unistd.h>

#include <cstring>
#include <iostream>

#include "Logger.hpp"

/**
 * @brief Constructs a POSIX TCP transport with an existing socket.
 * @param socket_fd The file descriptor of the connected socket.
 */
TcpTransport::TcpTransport(int socket_fd) : fd(socket_fd) {}

/**
 * @brief Destructor ensures that the socket is closed.
 */
TcpTransport::~TcpTransport() {
    close();
}

void TcpTransport::setCloseCallback(CloseCallback onClose) {
    closeCallback_ = onClose;
}

/**
 * @brief Starts receiving data on the transport.
 * @param onReceive Callback function to handle received data.
 */
void TcpTransport::start(ReceiveCallback onReceive) {
    // Required since nothing prevents start() from being called twice for the same instance.
    if (running.exchange(true))
        return;

    auto& logger = Logger::instance();
    logger.trace("Starting reader thread for fd " + std::to_string(fd));

    readerThread = std::jthread([this, onReceive](std::stop_token st) {
        auto& logger = Logger::instance();
        logger.trace("Reader thread started for fd " + std::to_string(fd));

        char buffer[1024];
        bool connection_closed_by_peer = false;

        while (!st.stop_requested() && running.load()) {
            logger.trace("Calling read() on fd " + std::to_string(fd));
            ssize_t n = read(fd, buffer, sizeof(buffer));
            logger.trace("read() returned " + std::to_string(n) + " for fd " + std::to_string(fd));

            if (n <= 0) {
                if (n == 0) {
                    logger.trace("Client disconnected (EOF) on fd " + std::to_string(fd));
                    connection_closed_by_peer = true;
                } else {
                    logger.error("Read error on fd " + std::to_string(fd) + ": " +
                                 std::string(strerror(errno)));
                    connection_closed_by_peer = true;
                }
                running = false;
                break;
            }

            std::string payload(buffer, n);
            onReceive(payload);
        }

        logger.trace("Reader thread EXITING for fd " + std::to_string(fd));

        // Notify session that connection died
        if (connection_closed_by_peer && closeCallback_) {
            logger.trace("Invoking close callback for fd " + std::to_string(fd));
            closeCallback_();
        }
    });
}

/**
 * @brief Sends data over the transport.
 * @param data The data to send.
 * @return Number of bytes sent, or -1 on error.
 */
void TcpTransport::send(const std::string& data) {
    if (!running.load()) {
        return;
    }

    ssize_t sent = write(fd, data.data(), data.size());
    if (sent < 0) {
        auto& logger = Logger::instance();
        logger.error("Write error on fd " + std::to_string(fd) + ": " +
                     std::string(strerror(errno)));
        running = false;
    }
}

/**
 * @brief Closes the TCP socket and terminates the reading loop.
 */
void TcpTransport::close() {
    // Required to be consistent with the similar check in start().
    if (!running.exchange(false))
        return;

    auto& logger = Logger::instance();
    logger.debug("Closing transport on fd " + std::to_string(fd));

    // Close socket first to unblock read()
    if (fd >= 0) {
        shutdown(fd, SHUT_RDWR);
        ::close(fd);
        fd = -1;
    }

    logger.debug("Transport closed");
}
